intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py

```diff
@@ -0,0 +1,271 @@
+            continue
+        xs.append(i)
+        ys.append(math.log(sum(ppls) / len(ppls)))  # average over samples
+    ax.plot(xs, ys, label=label)
+
+    xmax = xs[torch.argmax(torch.tensor(ys)).item()]
+    ymax = max(ys)
+    return xmax, ymax, xs, ys
+
+
+def draw_ppl(img_path: str, ppl_data: List[Dict[int, float]], ctx_size: int, model_title: str):
+    fig, ax = plt.subplots()
+    xmax, ymax, _, _ = add_log_ppl_line(ax, ppl_data)
+    ax.annotate(f"max={ymax:.4f}", (xmax, ymax))
+
+    ctx_line = ax.axvline(ctx_size, linestyle='--', color='r')
+    ctx_line.set_label('KV Cache Size')
+    ax.set_xlabel('Context Length')
+    ax.set_ylabel('Log Perplexity')
+    ax.legend()
+
+    ax.set_title(model_title)
+    fig.suptitle("Language modeling perplexity")
+    fig.savefig(img_path)
+
+    print(f"Max PPL: {math.exp(ymax)}")
+    return fig
+
+
+def add_quant_args(parser: argparse.ArgumentParser):
+    group = parser.add_argument_group('quantize config')
+    group.add_argument('--use_cache',
+                       action="store_true",
+                       help="Use local quantized model if file exists")
+    group.add_argument(
+        "--weight_dtype",
+        choices=["int4", "int8"],
+        help="Data type of quantized weight: int4/int8 (default: int4)",
+        default="int4",
+    )
+    group.add_argument(
+        "--alg",
+        type=str,
+        help="Quantization algorithm to use: sym/asym (default: sym)",
+        default="sym",
+    )
+    group.add_argument(
+        "--group_size", type=int, help="Group size: Int (default: 32)", default=32
+    )
+    group.add_argument(
+        "--scale_dtype",
+        type=str,
+        help="Data type of scales: bf16/fp32 (default: fp32)",
+        default="fp32",
+    )
+    group.add_argument(
+        "--compute_dtype",
+        type=str,
+        help="Data type of Gemm computation: int8/bf16/fp32 (default: int8)",
+        default="int8",
+    )
+    group.add_argument(
+        "--use_ggml",
+        action="store_true",
+        help="enable ggml for quantization and inference",
+    )
+    return group
+
+
+def add_run_args(parser: argparse.ArgumentParser):
+    group = parser.add_argument_group('model run config')
+    group.add_argument(
+        "--n_keep",
+        type=int,
+        help="Number of tokens to keep from the initial prompt: Int (default: 0; -1 = all)",
+        default=1,
+    )
+    group.add_argument(
+        "--shift_roped_k",
+        action="store_true",
+        help="Use ring-buffer and thus do not re-computing after reaching ctx_size (default: False)",
+    )
+    group.add_argument("--memory_dtype",
+                       type=str,
+                       help="Data type of the kv memory",
+                       choices=['f32', 'f16', 'auto'],
+                       default="auto")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Evaluate perplexity for a model givan a dataset")
```

Comment:
can we reuse the run.py?

Reply:
I tried but failed. There are some args in run.py (e.g. --color --prompt --build_dir) that does not make sense to ppl while there are some args (e.g. --dataset_name --n_pred_per_sample  --data_text_concat --fig_path) for ppl does not make sense to run.py.

Reply:
It is a good idea for me. But I guess there is enough python changes for this PR. Will try to generalize args in run/convert/inference/ppl to a `graph/scripts/utils.py` file